### PR TITLE
Feature 8 offline media

### LINKED ISSUE #8: Offline media
As a user I'd like the device to be able to offline media used in the surveys.

### PR DESCRIPTION
**Pull request by villejuutila**
Resolves #8

### CODE REVIEW

File: lib/files/offline_file_controller.dart [hunk 1]
```diff
@@ -0,0 +1,200 @@
+import "dart:async";
+import "dart:convert";
+import 'dart:io';
```
Comment (EVP1989) [inline lib/files/offline_file_controller.dart]:
```suggestion
import "dart:io";
```

File: lib/mqtt/listeners/surveys_listener.dart [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    } catch (e) {
+      logger.shout("Couldn't handle update survey message ${e.toString()}");
+    }
+  }
+
+  @override
+  void handleDelete(String message) async {
+    // try {
+      await OfflineFileController().getOfflineFile("https://staging-muisti-cdn.s3.eu-central-1.amazonaws.com/yleiset/2d61299a-3bf8-4a95-a7e3-a7ac05adfd7b");
+    //   String externalId = _getSurveyIdFromMessage(message);
```
Comment (EVP1989) [inline lib/mqtt/listeners/surveys_listener.dart]: Remove this if its not in use

**Review by EVP1989 — APPROVED** [review]
Approved with comment

File: lib/mqtt/listeners/abstract_listener.dart [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+/// Abstract MQTT Listener class
+abstract class AbstractMqttListener {
+
+  abstract String baseTopic;
+  
+  /// Returns a map of topic:callback entries handled by this listener.
+  Map<String, Function(String)> getListeners() => {
+    "$baseTopic/create":handleCreate,
+    "$baseTopic/update":handleUpdate,
+    "$baseTopic/delete":handleDelete
```
Comment (lauriLukkarinen) [inline lib/mqtt/listeners/abstract_listener.dart]:
```suggestion
    "$baseTopic/create": handleCreate,
    "$baseTopic/update": handleUpdate,
    "$baseTopic/delete": handleDelete
```

File: lib/mqtt/mqtt_client.dart [hunk 4]
```diff
@@ -54,11 +56,15 @@ class MqttClient {
       if (c.isEmpty) {
         return;
       }
-      final MqttPublishMessage message = c[0].payload as MqttPublishMessage;
-      final payload =
-      MqttPublishPayload.bytesToStringAsString(message.payload.message);
 
-      logger.info("Received message:$payload from topic: ${c[0].topic}>");
+      if (listeners.containsKey(c[0].topic)) {
```
Comment (lauriLukkarinen) [inline lib/mqtt/mqtt_client.dart]: Maybe 'messages' instead of 'c'?

**Review by lauriLukkarinen — CHANGES_REQUESTED** [review]

File: lib/files/offline_file_controller.dart (hunk 1 above)
Comment (villejuutila) [reply lib/files/offline_file_controller.dart]: Resolved

File: lib/mqtt/listeners/surveys_listener.dart (hunk 2 above)
Comment (villejuutila) [reply lib/mqtt/listeners/surveys_listener.dart]: Resolved

File: lib/mqtt/listeners/abstract_listener.dart (hunk 3 above)
Comment (villejuutila) [reply lib/mqtt/listeners/abstract_listener.dart]: Resolved

File: lib/mqtt/mqtt_client.dart (hunk 4 above)
Comment (villejuutila) [reply lib/mqtt/mqtt_client.dart]: Resolved

**Review by lauriLukkarinen — APPROVED** [review]

**Pull request merged**